packages/blocks/src/components/plain-text/plain-text.ts

```diff
@@ -0,0 +1,185 @@
+              rangeRect.left +
+              rangeRect.width -
+              (thisRect.left + thisRect.width);
+            moveX = Math.max(this._lastScrollLeft, moveX);
+          }
+
+          this.scrollLeft = moveX;
+        });
+      })
+    );
+
+    assertExists(this.virgoContainer);
+    vEditor.mount(this.virgoContainer);
+    vEditor.setReadonly(this.readonly);
+  }
+
+  private _onStackItemAdded = (event: { stackItem: PlainTextStackItem }) => {
+    const vRange = this.vEditor?.getVRange();
+    if (vRange) {
+      event.stackItem.meta.set('plaintext-v-range', vRange);
+    }
+  };
+
+  private _onStackItemPopped = (event: { stackItem: PlainTextStackItem }) => {
+    const vRange = event.stackItem.meta.get('plaintext-v-range');
+    if (vRange && this.vEditor?.isVRangeValid(vRange)) {
+      this.vEditor?.setVRange(vRange);
+    }
+  };
+
+  private _unmount() {
+    if (this.vEditor?.mounted) {
+      this.vEditor.unmount();
+    }
+    this._vEditor = null;
+  }
+
+  override async getUpdateComplete(): Promise<boolean> {
+    const result = await super.getUpdateComplete();
+    await this.vEditor?.waitForUpdate();
+    return result;
+  }
+
+  override connectedCallback() {
+    super.connectedCallback();
+
+    assertExists(this.yText, 'plain-text need yText to init.');
+    assertExists(this.yText.doc, 'yText should be bind to yDoc.');
+
+    // Plain-Text controls undo-redo itself if undoManager is not provided
+    if (!this.undoManager) {
+      this.undoManager = new Workspace.Y.UndoManager(this.yText, {
+        trackedOrigins: new Set([this.yText.doc.clientID]),
+      });
+
+      this.disposables.addFromEvent(this, 'keydown', (e: KeyboardEvent) => {
+        if (e.ctrlKey || e.metaKey) {
+          if (e.key === 'z') {
+            if (e.shiftKey) {
+              this.undoManager.redo();
+            } else {
+              this.undoManager.undo();
+            }
+            e.stopPropagation();
+          }
+        }
+      });
+
+      this.undoManager.on('stack-item-added', this._onStackItemAdded);
+      this.undoManager.on('stack-item-popped', this._onStackItemPopped);
+      this.disposables.add({
+        dispose: () => {
+          this.undoManager.off('stack-item-added', this._onStackItemAdded);
+          this.undoManager.off('stack-item-popped', this._onStackItemPopped);
+        },
+      });
+    }
+
+    this.updateComplete.then(() => {
+      this._unmount();
+      this._init();
+
+      this.disposables.add({
+        dispose: () => {
+          this._unmount();
+        },
+      });
+    });
+
+    this.disposables.addFromEvent(this, 'scroll', () => {
+      this._lastScrollLeft = this.scrollLeft;
+    });
+  }
+
+  override updated() {
+    if (this._vEditor) {
+      this._vEditor.setReadonly(this.readonly);
+    }
+  }
+
+  override render() {
+    return html`<div class="virgo-container"></div>`;
+  }
+}
+
+declare global {
+  interface HTMLElementTagNameMap {
+    'plain-text': PlainText;
```

Comment:
Since all custom elements share the same namespace, it is suggested to add a prefix to prevent unexpected name conflicts with application developers.

Reply:
It may need to add prefix for all blocksuite components. (worth creating a separate issue)